server/queue/queue-api/src/test/java/org/apache/james/queue/api/PriorityMailQueueContract.java

```diff
@@ -0,0 +1,225 @@
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.Date;
+import java.util.stream.IntStream;
+
+import javax.mail.MessagingException;
+
+import org.apache.mailet.Mail;
+import org.apache.mailet.base.MailAddressFixture;
+import org.apache.mailet.base.test.FakeMail;
+import org.junit.jupiter.api.Test;
+
+import com.github.fge.lambdas.Throwing;
+import com.github.steveash.guavate.Guavate;
+import com.google.common.collect.ImmutableList;
+
+public interface PriorityMailQueueContract {
+
+    MailQueue getMailQueue();
+
+    @Test
+    default void priorityShouldReorderMailsWhenDequeing() throws Exception {
+        getMailQueue().enQueue(mailBuilder()
+            .name(NAME3)
+            .attribute(MailPrioritySupport.MAIL_PRIORITY, 3)
+            .build());
+
+        getMailQueue().enQueue(mailBuilder()
+            .name(NAME9)
+            .attribute(MailPrioritySupport.MAIL_PRIORITY, 9)
+            .build());
+
+        getMailQueue().enQueue(mailBuilder()
+            .name(NAME1)
+            .attribute(MailPrioritySupport.MAIL_PRIORITY, 1)
+            .build());
+
+        getMailQueue().enQueue(mailBuilder()
+            .name(NAME8)
+            .attribute(MailPrioritySupport.MAIL_PRIORITY, 8)
+            .build());
+
+        getMailQueue().enQueue(mailBuilder()
+            .name(NAME6)
+            .attribute(MailPrioritySupport.MAIL_PRIORITY, 6)
+            .build());
+
+        getMailQueue().enQueue(mailBuilder()
+            .name(NAME0)
+            .attribute(MailPrioritySupport.MAIL_PRIORITY, 0)
+            .build());
+
+        getMailQueue().enQueue(mailBuilder()
+            .name(NAME7)
+            .attribute(MailPrioritySupport.MAIL_PRIORITY, 7)
+            .build());
+
+        getMailQueue().enQueue(mailBuilder()
+            .name(NAME4)
+            .attribute(MailPrioritySupport.MAIL_PRIORITY, 4)
+            .build());
+
+        getMailQueue().enQueue(mailBuilder()
+            .name(NAME2)
+            .attribute(MailPrioritySupport.MAIL_PRIORITY, 2)
+            .build());
+
+        getMailQueue().enQueue(mailBuilder()
+            .name(NAME5)
+            .attribute(MailPrioritySupport.MAIL_PRIORITY, 5)
+            .build());
+
+        ImmutableList<MailQueue.MailQueueItem> items = IntStream.range(1, 11).boxed()
+            .map(Throwing.function(i -> {
+                MailQueue.MailQueueItem item1 = getMailQueue().deQueue();
+                item1.done(true);
+                return item1;
+            }))
+            .collect(Guavate.toImmutableList());
+
+        assertThat(items)
+            .extracting(MailQueue.MailQueueItem::getMail)
+            .extracting(Mail::getName)
+            .containsExactly(NAME9, NAME8, NAME7, NAME6, NAME5, NAME4, NAME3, NAME2, NAME1, NAME0);
+    }
+
+    @Test
+    default void negativePriorityShouldBeConsideredZero() throws Exception {
+        getMailQueue().enQueue(mailBuilder()
+            .name(NAME2)
```

Comment:
not NAME0 ?

Reply:
You are right, I made a mistake while testing these edge cases. I's not considered zero but  as a normal priority.

I will rework this. Sorry for the mistake.